crates/subspace-farmer/src/plot.rs

```diff
@@ -202,6 +202,33 @@ impl Plot {
         })
     }
 
+    /// Creates a new plot from any kind of plot file
+    pub fn with_plot_file(
+        plot: impl PlotFile + Send + 'static,
+        base_directory: impl AsRef<Path>,
+        address: PublicKey,
+        max_piece_count: u64,
+    ) -> Result<Plot, PlotError> {
```

Comment:
This is better to read IMO:
```suggestion
    pub fn with_plot_file<P>(
        plot: P,
        base_directory: impl AsRef<Path>,
        address: PublicKey,
        max_piece_count: u64,
    ) -> Result<Plot, PlotError>
    where
        P: PlotFile + Send + 'static {
```

And you don't have to always call generic `T`. It only makes sense when generic can be really anything, so being more specific and use `P` or `PF` helps with reading.

Reply:
What is the rule of thumb here though? How do you choose generics over impl traits?

Reply:
We almost never use imp traits, I guess that is a rule of thumb in our codebase

Reply:
I guess we can use impl traits as long as it is only 1 trait there right?

Reply:
But why? It is typically more noisy and quickly prevents from basic function signature fitting one line.

Reply:
Don't you find more useful to look at types? With generics, it is always that you need to read all the arguments of the function and only after that see their types. Impl traits are much easier to scan in that sense IMO.

Also impl traits can't have dependencies between each other like generics can. Here I mean dependencies like this:
```rust
fn from_iterator<T, I>(a: I) -> Vec<T>
where
    I: Iterator<Item = T>;
```
Here `I` depends on `T`. With that in head it also makes it easier to scan through impl traits.

Reply:
We fairly regularly have generics that end up looking like this: https://github.com/subspace/subspace/blob/c3c40b74d4ec2302c577d709f2873bff6d642576/cumulus/client/cirrus-executor/src/lib.rs#L156-L182

Approach with `where` scales gracefully to that level. It seems beneficial to see that the function is generic right from the name and being able to forward other generics down when type inference doesn't work without creating temporary variables: https://github.com/subspace/subspace/blob/d2d87be5503bf19a09b0f4976ebeab047a6b3530/crates/sc-consensus-subspace/src/lib.rs#L709